### DISCUSSION TITLE
Bad value in VIRTUAL_PATH causes template failure, which causes container to exit

### DISCUSSION THREAD

**Discussion by hartzell** [nginx-proxy/nginx-proxy#2123] [Q&A]
I'm using nginx-proxy with a custom template to map ports from docker containers to paths on the front end via a `VIRTUAL_PATH` environment variable that specifies a path on the front end and the port inside the container.

E.g., I might start up a container that has Jenkins running inside that's listening on port 8080.  I'd like that visible at `/moose/jenkinses/foo`.  I'd specify `VIRTUAL_PATH=/moose/jenkinses/foo:8080`.

Here's the bit of the template that does the work (the rest is stolen from your examples):

```
server {
  server_name _; # This is just an invalid value which will never trigger on a real hostname.
  listen 80;
  access_log /var/log/nginx/access.log vhost;

  {{ range $port_path, $containers := groupByMulti $ "Env.VIRTUAL_PATH" "," }}
    {{ $info := split $port_path ":" }}
    location {{ index $info 0 }} {
      {{ range $index, $value := $containers }}
        {{ $network := index $value.Networks 0 }}
        proxy_pass http://{{ $network.IP }}:{{ index $info 1 }};
      {{ end }}
    }
  {{ end }}
}
```

If "someone" screws up and provides an `VIRTUAL_PATH` that only contains a hostname (they forget the port) then the `{{ index $info 1 }}` reference causes an error and the container ends up exiting.

```
dockergen.1 | 2017/04/13 16:20:23 Template error: template: nginx.tmpl:73:47: executing "nginx.tmpl" at <index $info 1>: error calling index: index out of range: 1
```

### Questions:

Ideally I'd like to continue proxying the other containers that are running and just report an error for the problematic container.

1. Is there something I can do in the template that will make this more robust (I'm a go template neophyte)?

2. Is there something that docker-gen or nginx-proxy can do to be more robust in cases like this?

**Comment by thomasleveil** [comment]
docker-gen fills up naively a template, there is no warranty that the produced configuration will work if the data provided to the template is broken. You might find that a tool such as [Træfik](https://traefik.io/) is more suited for a production environment.

**Comment by hartzell** [comment]
I started off with Traefik but found docker-gen easier to understand.  Perhaps I need to switch back.

My problem is not that the resulting configuration doesn't work, it's that the error in the template expansion causes docker-gen to exit and the entire container goes away.

I wonder if there's a way to check that the `split` I do actually produces a pair of values in `$info` so that the template doesn't try to access a non-existent value?

**Comment by thomasleveil** [comment]
That's the pain of being limited to the few functions the template syntax offers. Looking at the [functions](https://github.com/jwilder/docker-gen/blob/master/README.md#functions) provided by docker-gen, I can't find a way to count elements :/

**Comment by hartzell** [comment]
Would it be safer for docker-gen to refuse to replace the previous configuration if the template generation produces an error?
